Tests: Move Signup tests to single test runner

stephanethomas opened the pull request:
This pull request is part of https://github.com/Automattic/wp-calypso/issues/3942. It updates the unit test suites of the signup components to use the `client` test runner:

##### `client/lib/signup`

```
$ make test

  SignupDependencyStore
    ✓ should return an empty object at first
    ✓ should not store dependencies if none are included in an action
    ✓ should store dependencies if they are provided in either signup action

  SignupFlowController
    controlling a simple flow
      ✓ should run the onComplete callback with the flow destination when the flow is completed
    controlling a flow w/ an asynchronous step
      ✓ should call apiRequestFunction on steps with that property
      ✓ should not call apiRequestFunction multiple times
    controlling a flow w/ dependencies
      ✓ should call the apiRequestFunction callback with its dependencies
      ✓ should throw an error when the flow is completed without all dependencies provided
    controlling a flow w/ a delayed step
      ✓ should submit steps with the delayApiRequestUntilComplete once the flow is complete
      ✓ should not submit delayed steps if some steps are in-progress

  SignupProgressStore
    ✓ should return an empty at first
    ✓ should store a new step
    ✓ should not store the same step twice
    ✓ should not be possible to mutate
    ✓ should store multiple steps in order
    ✓ should mark submitted steps without an API request method as completed
    ✓ should mark submitted steps with an API request method as pending
    ✓ should mark only new saved steps as in-progress
    ✓ should set the status of a signup step
    timestamps
      ✓ should be updated at each step

  20 passing (223ms)
```

##### `client/signup`

```
$ make test

  flows.js
    ✓ should return the full flow when the user is not logged in
    ✓ should remove the user step from the flow when the user is not logged in

  utils.getLocale
    ✓ should find the locale anywhere in the params
    ✓ should return undefined if no locale is present in the params

  utils.getStepName
    ✓ should find the step name in either the stepName or flowName fragment
    ✓ should return undefined if no step name is found

  utils.getFlowName
    ✓ should find the flow name in the flowName fragment if present
    ✓ should return the default flow if the flow is missing
    ✓ should return the result of filterFlowName if it is a function and the flow is missing
    ✓ should return the result of filterFlowName if it is a function and the flow is not valid
    ✓ should return the result of filterFlowName if it is a function and the requested flow is present
    ✓ should return the passed flow if the result of filterFlowName is not valid
    ✓ should call filterFlowName with the default flow if it is a function and the flow is not valid
    ✓ should call filterFlowName with the requested flow if it is a function and the flow is valid

  utils.getValidPath
    ✓ should redirect to the default if no flow is present
    ✓ should redirect to the current flow default if no step is present
    ✓ should redirect to the default flow if the flow is the default
    ✓ should redirect invalid steps to the default flow if no flow is present
    ✓ should preserve a valid locale to the default flow if one is specified
    ✓ should redirect invalid steps to the current flow default
    ✓ should preserve a valid locale if one is specified
    ✓ should handle arbitrary step section names
    ✓ should handle arbitrary step section names in the default flow

  utils.getValueFromProgressStore
    ✓ should return the value of the field if it exists
    ✓ should return null if the field is not present

  utils.mergeFormWithValue
    ✓ should return the form with the field added if the field doesn't have a value
    ✓ should return the form unchanged if there is already a value in the form


  27 passing (263ms)
```

##### `client/signup/config`

```
$ make test

  signup/config
    ✓ should not have overlapping step/flow names

  1 passing (335ms)
```

#### Testing instructions
 
1. Run `git checkout update/signup-tests`
2. Execute `make test` and check that all tests for each module pass

blowery commented:
@stephanethomas I took the liberty of fixing up the rest of the tests since the way forward may not have been obvious. Does what I did make sense?

gziolo commented:
It's quite strange, but it look like test not related to this PR failed:
[should interpret a shortcode with content](https://github.com/Automattic/wp-calypso/blob/master/client/lib/shortcode/test/index.js#L251)

Maybe we need to move rewire call inside `describe` block?

gziolo commented:
Another reason could be:
`global.localStorage = require( 'localStorage' );`
on the top of all test files.

stephanethomas commented:
Thank you very much to both of you! It would have taken me much more time without your help.

I've updated two more test suites, and rebased. This is ready for review.

gziolo commented:
CircleCI bombs. It looks like one Makefile wasn't removed:
> Error: Cannot find module 'test/helpers/use-filesystem-mocks'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at Suite.<anonymous> (/home/ubuntu/wp-calypso/client/signup/test/flows.js:9:2)
    at context.describe.context.context (/home/ubuntu/wp-calypso/node_modules/mocha/lib/interfaces/bdd.js:47:10)
    at Object.<anonymous> (/home/ubuntu/wp-calypso/client/signup/test/flows.js:6:1)
    at Module._compile (module.js:410:26) ........

Also there is another error which is something most likely related to test runner itself:

> Error: Cannot find module 'signup/test/index'
    at Function.Module._resolveFilename (module.js:326:15)
    at Function.Module._load (module.js:277:25)
    at Module.require (module.js:354:17)
    at require (internal/module.js:12:17)
    at /home/ubuntu/wp-calypso/test/load-suite.js:8:38
    at Array.forEach (native)
    at /home/ubuntu/wp-calypso/test/load-suite.js:8:17
    at Array.forEach (native)
    at requireTestFiles (/home/ubuntu/wp-calypso/test/load-suite.js:4:24)
    at Suite.<anonymous> (/home/ubuntu/wp-calypso/test/load-suite.js:11:5)

This is because Circle CI recognises `client/signup/test/lib/abtest/index.js` as test file ...

gziolo commented:
Hehe, it looks like CircleCi wants to load `client/signup/test/lib/abtest/index.js`.

Most likely Circle CI uses Ruby `glob` together with with Ruby `convertGlobToRegEx`. See https://crossclj.info/ns/circleci/circle-util/0.1.1/circle-util.fs.html#_glob-matcher-fn.
In effect it translates `client/**/test/*.js` into `/^client\/.*.*\/test\/.*\.js$/` which would explain why it doesn't work as expected.
@blowery and ideas how to fix that?

gziolo commented on client/tests.json:
```diff
@@ -55,7 +55,7 @@
 		"cart-values": {
```
I used reformat code for this file. Sorry about that.

blowery commented:
> Hehe, it looks like CircleCi wants to load client/signup/test/lib/abtest/index.js

I think it's because `abtest` ends in `test`. :D It works fine for most other cases.

gziolo commented:
Yes, mockery solved this issue :)

blowery commented:
"solved" :smile: 

Someone else is going to run into this though. We should figure it out.

Two questions spring to mind:
- Why is that file making it past the whitelist?
- Why is the glob matching it?

blowery commented:
Commits on this are getting a bit tangled to. Should I try to rebase and squash to a sane set?

gziolo commented:
yes, please

Regarding glob. See my previous comment: https://github.com/Automattic/wp-calypso/pull/4301#issuecomment-201245881

gziolo commented:
> Why is that file making it past the whitelist?

I made simplified assumption that when there is `test` folder there are no more subfolders :) I didn't want to invest too much time into temporary solution. The truth is that it would pop anyway once we drop whitelist.

blowery commented:
> The truth is that it would pop anyway once we drop whitelist.

Yeah, good point. Which build made you think it's trying to load abtest/index as a test?

blowery commented:
> Most likely Circle CI uses Ruby glob together with with Ruby convertGlobToRegEx. See https://crossclj.info/ns/circleci/circle-util/0.1.1/circle-util.fs.html#_glob-matcher-fn.
In effect it translates client/**/test/*.js into /^client\/.*.*\/test\/.*\.js$/ which would explain why it doesn't work as expected.

What makes you think that's what the glob regexp is? `.*.*` doesn't really make sense.. If that's really how Ruby does a `**` glob, it's terribly broken.

gziolo commented:
> Which build made you think it's trying to load abtest/index as a test? I'm having a hard time finding it.

https://circleci.com/gh/Automattic/wp-calypso/11617
![screen shot 2016-03-25 at 13 41 38](https://cloud.githubusercontent.com/assets/699132/14044322/57c0545c-f28f-11e5-81c0-3a58bb882399.png)

> What makes you think that's what the glob regexp is? .*.* doesn't really make sense.. If that's really how Ruby does a ** glob, it's terribly broken.

I'm not 100% sure but it looks like Circle CI uses this implementation for glob: https://crossclj.info/ns/circleci/circle-util/0.1.1/circle-util.fs.html#_glob
Unless I googled something unrelated ;)

blowery commented:
Filed a bug with Circle, 4168

blowery commented on client/signup/config/test/index.js:
```diff
@@ -8,10 +8,22 @@ import isEmpty from 'lodash/isEmpty';
+describe( 'index', () => {
+	let flows, steps;
+
+	useFilesystemMocks( __dirname );
```
i wonder if useFilesystemMocks should be a thing we call from within useMockery? I think we're setting up a double-enable / double-cleanup here. It'll "work" but it's kinda fragile.

gziolo replied:
I forgot that filesystem mocks uses also mockery behind the scenes... Nice catch!

blowery commented on client/signup/config/test/index.js:
```diff
@@ -8,10 +8,22 @@ import isEmpty from 'lodash/isEmpty';
+
+	useFilesystemMocks( __dirname );
+
+	useMockery( ( mockery ) => {
```
or we could convert this to a `before` and import mockery directly and rely on useFilesystemMocks to cleanup. That feels like relying on incidental behavior though.

blowery commented:
I think this is fine to merge. We can circle back and clean up the mockery usage later. That's going to be a bigger PR anyway.

gziolo commented:
I added this bullet point to #3822:
- Make sure `useFilesystemMocks` and `useMockery` work properly together

gziolo commented:
Let's :shipit:

stephanethomas commented:
And I thought updating these tests would be easy :blush:.

I'm going to rebase and clean the commit history. I'll merge it afterwards.

blowery commented:
> And I thought updating these tests would be easy :blush:.

Yeah, anything that uses mocks is ... interesting. :)

gziolo commented:
thanks @stephanethomas for cleaning up commits history :)

stephanethomas commented:
Well, thank you to @blowery and you for doing the hard work :smiley:.

blowery commented:
> Filed a bug with Circle, 4618

@gziolo Turns out we saw this because this particular code branch was using a glob with slashes in it for client tests. See https://circleci.com/gh/Automattic/wp-calypso/11617#config

gziolo commented:
There are also Circle CI builds for this PR failed also when without slashes. I did this change during my investigation. I could send link to earlier build run.

Dnia 28.03.2016 o godz. 16:20 Ben Lowery <notifications@github.com> napisał(a):

> Filed a bug with Circle, 4618
> 
> @gziolo Turns out we saw this because this particular code branch was using a glob with slashes in it for client tests. See https://circleci.com/gh/Automattic/wp-calypso/11617#config
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
>
